### PR TITLE
Fix uninstall + remove left over resources after test failure

### LINKED ISSUE #3599: Bug with the Velero uninstall cmd
There are some major and minor problems with the uninstall command: https://github.com/vmware-tanzu/velero/blob/main/pkg/cmd/cli/uninstall/uninstall.go#L42-L74

- At each operation, it returns if there is an API error. This has a high potential of leaving behind resources that are still in the cluster only because the code didn't get to the point where they are removed. The solution is to aggregate all (real) errors and return them only at the end.
- An API error is not necessarily an error in this context. Example: a namespace (or any other resource) not found returns an API error, but it should not be considered an actual error of the cmd. In this case a log message is sufficient. Other error types should be distinguished and returned/aggregated as an error.
- The [wait](https://github.com/vmware-tanzu/velero/blob/e9c997839ebbb04688a88ffb981eeb74de51b380/pkg/cmd/cli/uninstall/uninstall.go#L81) for the namespace to be terminated is a Kubernetes anti-pattern.  If this was added because of the e2e tests, then the test is the piece that needs to wait for the resource to be deleted before proceeding, not the cli cmd. Also, it's weird that the uninstall takes way longer than the install. :/ - If this was added for the benefit of the user, again, it's an anti pattern to make the user wait for a resource to change state but if we want to aid the user we can log a message of the ns phase (see below).
- Any new API calls should be added using the kubebuilder/controller runtime framework.

I'm addressing the code issues above + modifying e2e tests that rely on this uninstall.

This is what running `velero uninstall` twice in a row without any (actual) errors could look like:

```
❯ velero uninstall
Velero uninstalled

❯ velero uninstall
Velero installation namespace "velero" is terminating.
Velero installation rolebinding "velero" does not exist, skipping.
Velero CRDs do not exist, skipping.
Velero uninstalled
```

### PR DESCRIPTION
**Pull request by carlisia**
# Does your change fix a particular issue?

Fixes #3599

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required`.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: test/e2e/velero_utils.go [hunk 1]
```diff
@@ -79,52 +92,44 @@ func GetProviderVeleroInstallOptions(
...
 
 	if io.UseRestic {
 		fmt.Println("Waiting for Velero restic daemonset to be ready.")
-		if _, err = install.DaemonSetIsReady(factory, io.Namespace); err != nil {
+		if _, err = install.DaemonSetIsReady(client.DynamicFactory, io.Namespace); err != nil {
 			return errors.Wrap(err, errorMsg)
 		}
 	}
 
+	fmt.Printf("Velero is installed and ready to be tested in the %s namespace! ⛵ \n", io.Namespace)
```
Comment (ashish-amarnath) [inline test/e2e/velero_utils.go:126]: nit: "Velero is installed ..." is a duplicate message.

**Review by ashish-amarnath — CHANGES_REQUESTED** [review]
@carlisia I know that this PR is in draft state.
Considering that this is blocking the release, trying to reduce the latency here. So I made some comments. PTAL.

File: pkg/cmd/cli/uninstall/uninstall.go [hunk 2]
```diff
@@ -19,57 +19,126 @@ package uninstall
...
+		Use:   "uninstall",
+		Short: "Uninstall Velero",
+		Long: `Uninstall Velero along with the CRDs.
+
+The '--namespace' flag can be used to specify the namespace where velero is installed (default: velero).
+		`,
+		Example: `# velero uninstall -n backup`,
+		Run: func(c *cobra.Command, args []string) {
+
+			if !cli.GetConfirmation() {
```
[the remark this answers is not in the document]
Comment (dsu-igeek) [reply pkg/cmd/cli/uninstall/uninstall.go:57]: Let's add a --quiet or --force option so this is scriptable

File: test/e2e/kibishii_tests.go [hunk 3]
```diff
@@ -104,23 +119,23 @@ func RunKibishiiTests(client *kubernetes.Clientset, providerName, veleroCLI, vel
 		return errors.Wrap(err, "Failed to generate data")
 	}
 
-	if err := VeleroBackupNamespace(oneHourTimeout, veleroCLI, veleroNamespace, backupName, kibishiiNamespace, backupLocation, useVolumeSnapshots); err != nil {
-		VeleroBackupLogs(fiveMinTimeout, veleroCLI, veleroNamespace, backupName)
+	if err := veleroBackupNamespace(oneHourTimeout, veleroCLI, veleroNamespace, backupName, kibishiiNamespace, backupLocation, useVolumeSnapshots); err != nil {
```
Comment (dsu-igeek) [inline test/e2e/kibishii_tests.go:122]: These were all public so we didn't need to put all of the e2e tests in the same package.  Eventually I think we want to split this up.

File: pkg/cmd/cli/uninstall/uninstall.go [hunk 4]
```diff
@@ -19,57 +19,126 @@ package uninstall
...
+			fmt.Printf("Velero installation namespace %q is terminating.\n", namespace)
+		} else {
+			if err := kbClient.Delete(context.Background(), ns); err != nil {
+				errs = append(errs, errors.WithStack(err))
+			}
+		}
 	}
 
-	rolebinding := install.ClusterRoleBinding(veleroNamespace)
+	time.Sleep(time.Second * 60)
```
[the remark this answers is not in the document]
Comment (carlisia) [reply pkg/cmd/cli/uninstall/uninstall.go:96]: Yes, I not at all had the intention to leave this here :) - it needs to be user configurable.

File: test/e2e/kibishii_tests.go (hunk 3 above)
Comment (carlisia) [reply test/e2e/kibishii_tests.go:122]: When we do split these tests into separate packages will be a great time to export what is needed to be exported.

File: pkg/cmd/cli/uninstall/uninstall.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (carlisia) [reply pkg/cmd/cli/uninstall/uninstall.go:57]: Good call, thanks.

**Comment by carlisia** [comment]
Closing this issue in favor of braking this work into 3 separate PRs.

For reference:

PR to fix the uninstall issue: https://github.com/vmware-tanzu/velero/pull/3618

New issues to track the rest:
https://github.com/vmware-tanzu/velero/issues/3624 - e2e tests not cleaning up after failures
https://github.com/vmware-tanzu/velero/issues/3625 - Consolidate API clients for the e2e tests

**Pull request closed without merging**